packages/cells/src/widget.ts

```diff
@@ -1419,6 +1435,65 @@ export class MarkdownCell extends AttachmentsCell<IMarkdownCellModel> {
     return this._ready.promise;
   }
 
+  /**
+   * Text that represents the heading if cell is a heading.
+   * Returns empty string if not a heading.
+   */
+  get headingInfo(): { text: string; level: number } {
+    let text = this.model.value.text;
+    const lines = marked.lexer(text);
+    let line: any;
+    for (line of lines) {
+      if (line.type === 'heading') {
+        return { text: line.text, level: line.depth };
+      } else if (line.type === 'html') {
+        let match = line.raw.match(/<h([1-6])(.*?)>(.*?)<\/h\1>/);
```

Comment:
Making the groups lazy was enough to silence CodeQL? Interesting. I will leave a link to the earlier discussion: https://github.com/jupyterlab/jupyterlab/pull/10260#discussion_r640105771 just so it is easy to find it.